vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/HttpContext.java

```diff
@@ -0,0 +1,307 @@
+        responseFuture.complete(resp);
+      }
+    });
+    if (request.timeout > 0) {
+      req.setTimeout(request.timeout);
+    }
+    if (body != null) {
+      if (contentType != null) {
+        String prev = req.headers().get(HttpHeaders.CONTENT_TYPE);
+        if (prev == null) {
+          req.putHeader(HttpHeaders.CONTENT_TYPE, contentType);
+        } else {
+          contentType = prev;
+        }
+      }
+      if (body instanceof ReadStream<?>) {
+        ReadStream<Buffer> stream = (ReadStream<Buffer>) body;
+        if (request.headers == null || !request.headers.contains(HttpHeaders.CONTENT_LENGTH)) {
+          req.setChunked(true);
+        }
+        Pump pump = Pump.pump(stream, req);
+        stream.exceptionHandler(err -> {
+          req.reset();
+          if (!responseFuture.isComplete()) {
+            responseFuture.fail(err);
+          }
+        });
+        stream.endHandler(v -> {
+          pump.stop();
+          req.end();
+        });
+        pump.start();
+      } else {
+        Buffer buffer;
+        if (body instanceof Buffer) {
+          buffer = (Buffer) body;
+        } else if (body instanceof MultiMap) {
+          try {
+            MultiMap attributes = (MultiMap) body;
+            boolean multipart = "multipart/form-data".equals(contentType);
+            DefaultFullHttpRequest request = new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, io.netty.handler.codec.http.HttpMethod.POST, "/");
+            HttpPostRequestEncoder encoder = new HttpPostRequestEncoder(request, multipart);
+            for (Map.Entry<String, String> attribute : attributes) {
+              encoder.addBodyAttribute(attribute.getKey(), attribute.getValue());
+            }
+            encoder.finalizeRequest();
+            for (String headerName : request.headers().names()) {
+              req.putHeader(headerName, request.headers().get(headerName));
+            }
+            if (encoder.isChunked()) {
+              buffer = Buffer.buffer();
+              while (true) {
+                HttpContent chunk = encoder.readChunk(new UnpooledByteBufAllocator(false));
+                ByteBuf content = chunk.content();
+                if (content.readableBytes() == 0) {
+                  break;
+                }
+                buffer.appendBuffer(Buffer.buffer(content));
+              }
+            } else {
+              ByteBuf content = request.content();
+              buffer = Buffer.buffer(content);
+            }
+          } catch (Exception e) {
+            throw new VertxException(e);
+          }
+        } else if (body instanceof JsonObject) {
+          buffer = Buffer.buffer(((JsonObject)body).encode());
+        } else {
+          buffer = Buffer.buffer(Json.encode(body));
+        }
+        req.end(buffer);
+      }
+    } else {
+      req.end();
+    }
+  }
+
+  public <T> T get(String key) {
+    return getOrDefault(key, null);
+  }
+
+  public <T> T getOrDefault(String key, T def) {
```

Comment:
A `getOrDefault(String, Supplier)` would be useful to avoid creating default values again and again in the interceptor.